includes/class-amp-theme-support.php

```diff
@@ -417,15 +419,26 @@ public static function finish_init() {
 
 		self::ensure_proper_amp_location();
 
-		$theme_support = self::get_theme_support_args();
+		$is_reader_mode = self::READER_MODE_SLUG === self::get_support_mode();
+		$theme_support  = self::get_theme_support_args();
 		if ( ! empty( $theme_support['template_dir'] ) ) {
 			self::add_amp_template_filters();
+		} elseif ( $is_reader_mode ) {
```

Comment:
Instead of having a separate condition for reader mode, it may be good to combine with `add_amp_template_filters`. The thinking is we may want to go ahead and try to load a template as if it is in Transitional mode, but then if we are about to load a `single.php` or `page.php` that contains `$this`, then we can at that point load it via `AMP_Post_Template`. This will avoid necessarily doing the current hard-coding of `single.php` and `page.php` as seen in:

https://github.com/ampproject/amp-wp/blob/451dd22cb6f80a631c95bec7fef1b0e02f8e9c48/includes/templates/class-amp-post-template.php#L198-L199

This would allow Reader templates to make use of the full template hierarchy, and ease transition to _Transitional_ templates.

In other words, if someone does:

```php
add_theme_support( 'amp', [ 'template_dir' => 'amp' ] );
```

This should either allow the templates in the `amp` directory to be classic `AMP_Post_Template`-based templates, or regular WordPress theme templates. We can start phasing out the use of `AMP_Post_Template`, only loading it when we detect it is needed.

In fact, there shouldn't be any real difference between Reader mode and Transitional mode when a `template_dir` is present.

Reply:
Instead of detecting `$this` inside of a template (which requires parsing the entire template file), we could also look into wrapping _all_ templates in a compatibility wrapper object by default, as regular WordPress templates don't use `$this`. However, I'm not 100% sure whether that cleanly works across global state (which is expected in regular WordPress templates).

Reply:
As opposed to parsing the entire template file, I was intending to just use a simple check to see if `'$this->'` occurs inside of the file. If so, it could then load it via the `AMP_Post_Template`. Otherwise, it could do normal `include`.

Reply:
You'll still have to read the entire file to find a potential `$this`. And this reading is separate from `include`ing the file, which you'll need to do as well, and which would normally happen via opcache and not hit the filesystem, so this adds more filesystem reads.

The compatibility wrapper I suggested would be read once and wrapped around each include. So most of the time, everything would still stay within the opcache. However, it adds complexity.

Reply:
Doesn't `file_get_contents()` use caching as well? Per [docs](https://www.php.net/manual/en/function.file-get-contents.php):

> `file_get_contents()` is the preferred way to read the contents of a file into a string. It will use memory mapping techniques if supported by your OS to enhance performance.

I'm not sure what these “memory mapping techniques” entail.

In any case, we're doing `file_get_contents()` for other files with each page load, for example:

https://github.com/ampproject/amp-wp/blob/c900d8484de1751bf56fd8b1c392154fa922eebe/templates/style.php#L55

Why would this be inherently different?

In core I see `get_post_embed_html()` uses `file_get_contents()` if `SCRIPT_DEBUG` is enabled. Also, `file_get_contents()` is used by `load_script_translations()` in core.

Reply:
In any case, migrating Reader mode templates to use a normal theme templates is beyond the scope of this PR.